Merge tensorboard summary and tensorboard output

CatherineSue opened the pull request:
Merge tensorboard_summary.Summary into 
tensorboard_output.TensorBoardOutput, which supports 
graph, scalar, tensor, and histogram logging to tensorboard.

The private function in TensorBoardOutput starts with '_'.
All other functions are public.

What this commit does:
    1. Optimize imports
    2. Merge class
    3. Solve the duplicate fields in custom scalars.(https://github.com/ryanjulian/rllab/pull/88#issuecomment-392245839)

Commit:
Merge tensorboard summary and tensorboard output

Merge tensorboard_summary.Summary into
tensorboard_output.TensorBoardOutput, which supports
graph, scalar, tensor, and histogram logging to
tensorboard.

The private function in TensorBoardOutput starts with '_'.
All other functions are public.

What this commit does:
1. Optimize imports
2. Merge class
3. Solve the duplicate fields in custom scalars.
(https://github.com/ryanjulian/rllab/pull/88#issuecomment-392245839)

jonashen commented on rllab/misc/tensorboard_output.py:
```diff
@@ -1,9 +1,14 @@
+import google.protobuf.json_format as json_format
 import numpy as np
 import tensorflow as tf
+from jsonmerge import merge
```
I think this needs to be alphabetical order?

CatherineSue replied:
like from imports before imports?
```
from jsonmerge import merge
...
import numpy as np
import tensorflow as tf
```

ryanjulian replied:
Please make sure all deps are listed in environment.yml

ryanjulian replied:
Alphabetize imports as if works like "from" or "import" or "as" are not there. only package/module/class/function names matter.

jonashen commented on rllab/misc/tensorboard_output.py:
```diff
@@ -41,12 +65,55 @@ def dump_tensorboard(self, step=None):
+        self._dump_histogram(run_step)
         self._dump_tensors()
 
+    def record_histogram(self, key, val):
```
Do you need a definition for this func

ryanjulian commented on rllab/misc/tensorboard_output.py:
```diff
@@ -1,9 +1,14 @@
+import shutil
+from os.path import dirname, abspath
 
+import google.protobuf.json_format as json_format
```
Please ensure all deps are in environment.yml

Even if a package happens to be accessibly because some other dependency uses it (transitive dependency), we need to make sure that environment.yml lists all dependencies in _our_ code.

ryanjulian commented on rllab/misc/logger.py:
```diff
@@ -1,26 +1,22 @@
-import base64
-
+import sys
 from contextlib import contextmanager
```
make sure these are all in alphabetical order.

ryanjulian reviewed (approved):
Looks great. Please address small comments and submit!

Commit:
Add deps in environment and alphabetize imports

Add dependencies jsonmerge and protobuf into environment.yml.
Alphabetize imports.

CatherineSue committed:
Merge branch 'integration' into tensor_op
